hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/RefreshMetaProcedure.java

```diff
@@ -0,0 +1,423 @@
+          puts.add(MetaTableAccessor.makePutFromRegionInfo(latestRegion));
+          LOG.debug("Adding a put to update region boundaries in meta: {}",
+            latestRegion.getRegionNameAsString());
+        }
+      }
+    }
+
+    if (!puts.isEmpty()) {
+      LOG.info("Adding/updating {} regions in meta table", puts.size());
+      executeBatchedPuts(connection, puts);
+    }
+
+    if (!deletes.isEmpty()) {
+      LOG.info("Removing {} regions from meta table", deletes.size());
+      executeBatchedDeletes(connection, deletes);
+    }
+
+    if (puts.isEmpty() && deletes.isEmpty()) {
+      LOG.info("No changes needed in meta table");
+    }
+  }
+
+  private void executeBatchedPuts(Connection connection, List<Put> puts) throws IOException {
+    for (int i = 0; i < puts.size(); i += CHUNK_SIZE) {
+      int end = Math.min(puts.size(), i + CHUNK_SIZE);
+      List<Put> chunk = puts.subList(i, end);
+
+      for (int attempt = 1; attempt <= 3; attempt++) {
+        try {
+          MetaTableAccessor.putsToMetaTable(connection, chunk);
+          LOG.debug("Successfully processed put batch {}-{}", i, end);
+          break;
+        } catch (IOException ioe) {
+          LOG.warn("Put batch {}-{} failed on attempt {}/3", i, end, attempt, ioe);
+          if (attempt == 3) {
+            throw ioe;
+          }
+          try {
+            Thread.sleep(100);
+          } catch (InterruptedException iex) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Interrupted during retry", iex);
+          }
+        }
+      }
+    }
+  }
+
+  private void executeBatchedDeletes(Connection connection, List<Delete> deletes) throws IOException {
+    for (int i = 0; i < deletes.size(); i += CHUNK_SIZE) {
+      int end = Math.min(deletes.size(), i + CHUNK_SIZE);
+      List<Delete> chunk = deletes.subList(i, end);
+
+      for (int attempt = 1; attempt <= 3; attempt++) {
+        try {
+          MetaTableAccessor.deleteFromMetaTable(connection, chunk);
+          LOG.debug("Successfully processed delete batch {}-{}", i, end);
+          break;
+        } catch (IOException e) {
+          LOG.warn("Delete batch {}-{} failed on attempt {}/3", i, end, attempt, e);
+          if (attempt == 3) {
+            throw e;
+          }
+          try {
+            Thread.sleep(100);
+          } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+            throw new IOException("Interrupted during retry", ie);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Determines if an update is needed by comparing current and latest regions.
+   */
+  boolean needsUpdate(List<RegionInfo> current, List<RegionInfo> latest) {
+    if (current == null || latest == null) {
+      LOG.warn("Cannot compare null region lists - current: {}, latest: {}",
+        current != null, latest != null);
+      return false;
+    }
```

Comment:
You've already done this check in the calling method. I think you should either do it here or there only.